mis/pipeline/pulse.py

```diff
@@ -20,59 +26,119 @@ class BasePulseShaper(ABC):
     """
-    Method that returns the correct PulseShaper based on configuration.
-    The correct pulse shaping method can be identified using the config, and an
-    object of this pulseshaper can be returned using this function.
 
-    Args:
-        instance (MISInstance): The MIS problem to embed.
-        config (Device): The quantum device to target.
+    def _get_interactions(
+        self, pos: np.ndarray, graph: nx.Graph, device: Device
+    ) -> tuple[list[float], list[float]]:
+        """Calculate the interaction strengths for connected and disconnected nodes.
 
-    Returns:
-        (BasePulseShaper): The representative Pulse Shaper object.
-    """
+        Args:
+            pos (np.ndarray): The position of the nodes.
+            graph (nx.Graph): The associated graph.
+            device (BaseDevice): Device used to calculate interaction coeff.
+
+        Returns:
+            tuple[list[float], list[float]]: Connected interactions, Disconnected interactions
+        """
 
-    return FirstPulseShaper(instance, config)
+        def calculate_edge_interaction(edge: tuple[int, int]) -> float:
+            pos_a, pos_b = pos[edge[0]], pos[edge[1]]
+            return float(device.interaction_coeff / (euclidean(pos_a, pos_b) ** 6))
+
+        connected = [calculate_edge_interaction(edge) for edge in graph.edges()]
+        disconnected = [calculate_edge_interaction(edge) for edge in nx.complement(graph).edges()]
+
+        return connected, disconnected
+
+    def _calc_bounds(self, reg: Register, graph: nx.Graph, device: Device) -> _Bounds:
+        _, disconnected = self._get_interactions(reg.register.sorted_coords, graph, device)
+        u_min, u_max = self._interaction_bounds(reg.register.sorted_coords, graph, device)
+        max_amp_device = device.channels["rydberg_global"].max_amp or np.inf
+        maximum_amplitude = min(max_amp_device, u_max + 0.8 * (u_min - u_max))
+
+        degree = graph.degree
+        assert isinstance(degree, DegreeView)
+        d_min = min(dict(degree).values())
+        d_max = max(dict(degree).values())
+        det_max_theory = (d_min / (d_min + 1)) * u_min
+        det_min_theory = sum(sorted(disconnected)[-d_max:])
+        det_final_theory = max([det_max_theory, det_min_theory])
+        det_max_device = device.channels["rydberg_global"].max_abs_detuning or np.inf
+        final_detuning = min(det_final_theory, det_max_device)
+
+        return _Bounds(maximum_amplitude=maximum_amplitude, final_detuning=final_detuning)
+
+    def _interaction_bounds(
+        self, pos: np.ndarray, graph: nx.Graph, device: Device
+    ) -> tuple[float, float]:
+        """Calculates U_min and U_max given the positions. It uses the edges of the
+        graph. U_min corresponds to minimal energy of two nodes connected in the
+        graph. U_max corresponds to maximal energy of two nodes NOT connected in
+        the graph."""
+        connected, disconnected = self._get_interactions(pos, graph, device)
```

Comment:
In `_calc_bounds` you get these interaction strengths right before calling `_interaction_bounds`. Would it make sense to reuse them as opposed to recomputing them here ?

Reply:
I was planning to have another pass to document and optimize this. Do you mind if we handle this in a followup issue?

Reply:
Filed as #20 .